msyslog-v1.08a+smac/src/man/om_peo.8
.\"	$CoreSDI: om_peo.8,v 1.7.2.1.2.1.4.10 2001/11/20 09:56:23 alejo Exp $
.\"
.\" Copyright (c) 2001
.\"	Core-SDI SA. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of Core-SDI SA nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd May 10, 2000
.Dt OM_PEO 8
.Os Core-SDI
.Sh NAME
.Nm peo output module
.Nd
.Xr syslogd 8
output module used to protect log files
.Sh SYNOPSIS
.Nm peo
.Op Fl k Ar keyfile
.Op Fl l
.Op Fl m Ar hash_method
.Sh DESCRIPTION
.Nm peo
output module receives a message as an ascii string and calculates its 
.Em hash
key based on the last one generated for the previous message; the
module removes the last key and writes the new one into
.Ar keyfile .
This module's options are as follows:
.Bl -tag -width Ds
.It Fl k Ar keyfile
Specify the key file pathname; the default is
.Pa /var/ssyslog/.var.log.messages.key
.It Fl l
This option enables the line corrupted detection mode;
the module generates two keys, the first explained above and a second
key using a 
.Em mac
method based on two consecutive
.Em hash
functions, this new key is added into the 
.Em mac
file whose pathname is the same as
.Ar keyfile
with a ".mac" string added at the end (if this file does not exists,
is created automatically).
.It Fl m Ar hash_method
Specifies the hash method used to generate the key to put into the
.Ar keyfile, hash_method
should be one of 
.Cm md5, sha1,
or
.Cm rmd160;
the default is 
.Cm sha1.
.El
.Sh EXAMPLES
If you want to protect the
.Pa /var/log/authlog
file you should edit the
.Pa /etc/syslog.conf
file (see
.Xr syslog.conf 5
) and add a line with something like this:
.Pp
.Dl auth.info	%peo -l -k /var/ssyslog/.var.log.authlog.key %classic /var/log/authlog
.Pp
You should generate the initial key with
.Xr peochk 8
program, then rotate the logfile(s) and restart msyslog. Afterwards
you can check the logfile integrity with the same program.
.Sh SEE ALSO
.Li Vcr and Peo Revised documentation - http://www.corest.com/papers/peo.ps
.Xr syslog 3 ,
.Xr syslog.conf 5 ,
.Xr om_classic 8 ,
.Xr om_mysql 8 ,
.Xr om_pgsql 8 ,
.Xr om_regex 8 ,
.Xr om_tcp 8 ,
.Xr om_udp 8 ,
.Xr peochk 8 ,
.Xr syslogd 8
.Sh BUGS
.Bl -bullet
.It
Since the peo module is used to determine if a logfile is corrupted, care
must be taken on the configuration file, the following is not correct:
.Pp
.Dl *.err	/var/log/messages
.Pp
.Dl *.err	%peo -k /var/ssyslog/.var.log.messages.key
.Pp
the following is wrong either:
.Pp
.Dl *.err	%classic /var/log/messages
.Pp
.Dl *.err	%peo -k /var/ssyslog/.var.log.messages.key
.Pp
The correct line is:
.Pp
.Dl *.err	%classic /var/log/messages  %peo -k /var/ssyslog/.var.log.messages.key
.Pp
or
.Pp
.Dl *.err	%peo -k /var/ssyslog/.var.log.messages.key  %classic /var/log/messages
.It
Submit bugs at this project's Sourceforge Bug reporting system at:
http://sourceforge.net/tracker/?func=add&group_id=25741&atid=385117
You may also report them directly to the authors; send an email to
core.devel.alat@corest.com, describing the problem the most you can,
containing also machine description, hardware description, the
configuration file (/etc/syslog.conf), the OS description, and the
invoking command line.
The more you describe the bug, the faster we can fix it.
.El


html/ChangeLog.txt
$Id$

Feature Request:

Bugs:
1.  Change code to allow for tailing spaces when looking at the expression field

3/10/2004
	-  Removing any reference to company name

3/9/2004
	-  Changed out company logo
	-  Added GPL notice to all libraries

1/21/2004
	-  Found a bug where a host would email off an alert with no text.  I've put a stop gap fix in.

1/13/2004
        -  Fixed bug with msyslog.pgsql where SET STORAGE didn't have a space before it.
        -  Fixed bug where hour in time of rules would default to 18:00 for no good reason
	-  Fixed two problems with cloning rules:  1)  order was not preserved  2)  an imported version of the database couldn't clone because fields that were empty needed to have non-null defaults applied
	-  Forgot to merge in latest VPN reports

1/6/2004
        -  Added 'left menu' support for ACID and MRTG groups in security framework should the ever be added 8)

1/6/2004
	-  Added 'left menu' support for ACID and MRTG groups in security framework should the ever be added 8)

1/1/2004
	-  Fixed spelling error in with the word 'threshold' in the rules schema
	-  Fixed a bug in vacuumdb where it was 'ANALYZ' not 'ANALYZE'

12/19/2003
	-  Updated code to look into /opt/apache

12/10/2003
	-  Updated database to schema to not use compressed text fields(we'll see how this performs)
	-  working on adding interfaces to more lock data and other new stats with PostgreSQL V7.4
	-  cleaned up some button descriptions on the maintenance page, also added a lock view as well as a settings view

12/9/2003
	-  Finished rule.php support for basic timer maintenance.  Need to add another page to graft timers onto rules.
	-  processlogs.php is now setup to support rule timers
	-  need to write rule expiration process 
	-  started converting away from compressed text in hopes of providing faster data retrieval.... also pulled OIDs from DB definition

12/8/2003
	-  Adding support for date and time based rules with date ranges, day of week selections, as well as deleted rules, need to add interface to control rule timestamp properties

12/7/2003
	-  finished adding accumulation thresholds

12/3/2003
	-  adding support for both types of thresholds.  Need to update web pages to reflect new radio buttons.
	-  added web configuration support for supression thresholds and accumulating thresholds, now onto updating the log processor
	-  basic supression works(kinda), need to verify functionality

12/2/2003
	-  customer profiles can now have multiple hosts added at once.
	-  fixed a stupid bug where <prev|refresh|next> didn't behave right.  next would stop working if you started at oct-26-2003 and it would stay on oct-26-2003
	-  Added web-based framework and database schema to support alert supression thresholds

12/1/2003
	-  Updated processlogs.php code to better deal with single entry/no rules vs mutliple rules

11/19/2003
	-  Updated mail table to enforce unique login ids(effectively stopping two processes from running at the same time, one will crash and die(safely))
	-  Updated openmail and closeopenmail to use transaction support since PostgreSQL no longer does server-side auto-commit(ie. convert everything to transactions)
	-  Transaction support should now be officially added, will do some testing

11/18/2003
	-  Updating program for support with PostGreSQL V7.4
	-  Fixed host process table to reflect the fact that multiple hosts are in there by default	

10/13/2003
	-  Added alert total to bottom of alert aggregation

10/6/2003
	-  finished support for alert aggregation
	-  updated 1stview to pull the current time and date
	-  fixed host.php bug where you could expire syslogs but not be forced to expire syslogs

10/5/2003
	-  still adding support for alert aggregation.  Basic aggregation works plus alert zooming but need to add support for across the board for other alert queries

10/3/2003
	-  added ability to un/suspend log processors from the web interface
	-  updated maintenance to rebuild all indexes in an better manner(ie. grab the index list from the DB rather then by hand)
	-  added additional framework to do alert aggregation interface.  Need to add 'aggregation code' for display

9/24/2003
	-  weeklyindexrebuild.php now pulls all indexes from the system and rebuilds them.  The result is that the system will now rebuild any new indexes without manual reconfiguration.

9/15/2003
	-  launchid was not initiated correctly in the clonerule.

9/2/2003
	-  Fixed a bug where '\'s at the end of a line caused problems because we were not properly dealing with them in general.  Fixed that.

8/23/2003	
	-  Created another bug when fixing 5000 line paging.  Timestamp was thrown off in view.php

8/18/2003	
	-  Syncing changes from production smt environment:  vacuumtsyslog.php
	-  Updated weeklyindexrebuild.php to account for the correct indexes
	-  Updated maint.php to account for the three new indexes for the launch program section

8/13/2003
	-  view.php has had several updates.  Paging should now be fixed.  Multiple searches appeared not to be working correctly.
	-  Needed to add lastid as hidden var if the variable was set
	-  Needed to use urlencode on top of htmlspecialcharacters, filters were broken because of it

8/11/2003
	-  1stfilter.php doesn't list 'global' filters that you down own
	-  modified view.php to not let the user save a filter with no description
	-  added support to delete all of a user's filters(ie. do before delete)
	-  another problem popped up with filter.php when I added the delete user filter option

8/1/2003	
	-  processlogs.php now supports launching external programs!

7/31/2003
	-  Updated vacuumdb.php to do a full vacuum of the TSyslog table.  Why?  Because the system doesn't reclaim disk space or use old delete space for some reason
	-  Almost finished adding launch program code, need to test.

7/27/2003
	-  Added weeklyindexrebuild.php which rebuilds all indexes at 5am Sunday morning

7/23/2003

msyslog-v1.08a+smac/src/modules/om_peo.c
/*	$CoreSDI: om_peo.c,v 1.41.2.5.4.17 2001/11/30 23:13:45 alejo Exp $	*/

/*
 * Copyright (c) 2001, Core SDI S.A., Argentina
 * All rights reserved
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither name of the Core SDI S.A. nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * om_peo -- peo autentication
 *
 * Author: Claudio Castiglia
 *
 */

#include "config.h"

#if TIME_WITH_SYS_TIME
# include <sys/time.h>
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/param.h>

#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <syslog.h>
#include <strings.h>

#include "../modules.h"
#include "../syslogd.h"
#if 1
#include "../peo/hash.h"
#else
extern char *default_keyfile;
#define	SHA1	0
#endif

#define MAXBUF	MAXSVLINE+MAXHOSTNAMELEN+20

struct om_peo_ctx {
	short	flags;
	int	size;
	int	hash_method;
	char   *keyfile;
	char   *macfile;
};

int
om_peo_write(struct filed *f, int flags, struct m_msg *msg, void *ctx)
{
	struct om_peo_ctx	*c;
	int			 fd, mfd, len, keylen, newkeylen;
	u_char			 key[41], mkey[41];
	unsigned char		 m[MAXBUF], newkey[41];
	char			 time_buf[16];

	dprintf(MSYSLOG_INFORMATIVE, "om_peo_write: Entering\n");
	
	if (f == NULL || ctx == NULL || msg == NULL)
		return (-1);

	c = (struct om_peo_ctx *) ctx;

	strftime(time_buf, sizeof(time_buf), "%b %e %H:%M:%S", &f->f_tm);
	time_buf[15] = '\0';
	len = snprintf((char *) m, MAXBUF, "%s %s %s\n", time_buf,
	    f->f_prevhost, msg->msg ? msg->msg : f->f_prevline) - 1;

	dprintf(MSYSLOG_INFORMATIVE, "om_peo_write: len = %i, msg->msg = %s\n ",
	    len, m);

	/* Open keyfile and read last key */
	if ( (fd = open(c->keyfile, O_RDWR, 0)) < 0) {
		dprintf(MSYSLOG_SERIOUS, "om_peo_write: opening keyfile: %s:"
		    " %s\n", c->keyfile, strerror(errno));
		return (-1);
	}
	bzero(key, sizeof(key));
	if ( (keylen = read(fd, key, 40)) < 0) {
		close(fd);
		dprintf(MSYSLOG_SERIOUS, "om_peo_write: reading form: %s:"
		    " %s\n", c->keyfile, strerror(errno));
		return (-1);
	}

	/* Open macfile and write mac'ed msg */
	if (c->macfile) {
		if ( (mfd = open(c->macfile, O_WRONLY, 0)) < 0) {
			close(fd);
			dprintf(MSYSLOG_SERIOUS, "om_peo_write: opening "
			    "macfile: %s: %s\n", c->macfile,
			    strerror(errno));
			return (-1);
		}
		lseek(mfd, (off_t) 0, SEEK_END);
		write(mfd, mkey, mac2(key, keylen, m, len, mkey));
		dprintf(MSYSLOG_INFORMATIVE, "om_peo_write: write to macfile"
		    " ok\n");
		close(mfd);
	}

	/* Generate new key and save it on keyfile */
	lseek(fd, (off_t)0, SEEK_SET);
	ftruncate(fd, (off_t)0);
	newkeylen = mac(c->hash_method, key, keylen, m, len, newkey);
	if (newkeylen == -1) {
		close(fd);
		dprintf(MSYSLOG_INFORMATIVE, "om_peo_write: generating "
		    "key[i+1]: keylen = %i: %s\n", newkeylen,
		    strerror(errno));
		return (-1);
	}
	write(fd, newkey, newkeylen);
	close(fd);
	return (1);
}


/*
 *  INIT -- Initialize om_peo
 *  args:
 * 
 *	-k <keyfile>		(default: /var/ssyslog/.var.log.messages)
 *	-l			line number corruption detect mode
 *				(generates a strcat(keyfile, ".mac") file)
 *	-m <hash_method>	md5, rmd160, or sha1 (default: sha1)
 *
 */
char		*keyfile;
char		*macfile;
void

release(void)
{
	if (keyfile != default_keyfile)
		free(keyfile);
	if (macfile != NULL)
		free(macfile);
}

int
om_peo_init(int argc, char **argv, struct filed *f, char *prog, void **ctx,
	    char **status)
{
	int			 ch;
	struct om_peo_ctx	*c;
	int			 hash_method;
	int			 mfd;
	char			 statbuf[2048];

	dprintf(MSYSLOG_INFORMATIVE, "om_peo_init: Entering, called by %s\n",
	    prog);
	
	if (argv == NULL || *argv == NULL || argc == 0 || f == NULL ||
	    ctx == NULL)
		return (-1);

	/* default values */
	hash_method = SHA1;
	keyfile = default_keyfile;
	macfile = NULL;
	mfd = 0;

	/* parse command line */
#ifdef HAVE_OPTRESET
	optreset = 1;
#endif
	optind = 1;
	while ( (ch = getopt(argc, argv, "k:lm:")) != -1) {
		switch (ch) {
		case 'k':
			/* set keyfile */
			release();
			if ( (keyfile = strdup(optarg)) == NULL)
				return (-1);
			break;
		case 'l':
			/* set macfile */
			mfd = 1;
			break;
		case 'm':
			/* set method */
			if ( (hash_method = gethash(optarg)) < 0) {
				release();
				errno = EINVAL;
				return (-1);
			}
			break;
		default:
			release();
			errno = EINVAL;
			return (-1);
		}
	}

	/* set macfile */
	if (mfd) {
		if ( (macfile = (char *) strmac(keyfile)) == NULL) {
			release();
			return (-1);
		}
		if (! (mfd = open(macfile, O_CREAT, S_IRUSR | S_IWUSR))) {
			if (errno != EEXIST) {
				release();
				return (-1);
			}
		} else
			close(mfd);
	}

	/* save data on context */
	if ( (c = (struct om_peo_ctx*)
	    calloc(1, sizeof(struct om_peo_ctx))) == NULL) {
		release();
		return (-1);
	}

	c->size = sizeof(struct om_peo_ctx);
	c->hash_method = hash_method;
	c->keyfile = keyfile; 
	c->macfile = macfile;
	*ctx = (void *) c;

	snprintf(statbuf, sizeof(statbuf), "om_peo: method: "
	    "%d\nkeyfile: %s\nmacfile: %s\n", hash_method, keyfile,
	    macfile);
	*status = strdup(statbuf);

	return (1);
}


int
om_peo_close(struct filed *f, void *ctx)
{
	struct om_peo_ctx *c;

	c = (struct om_peo_ctx *) ctx;
	dprintf(MSYSLOG_INFORMATIVE, "om_peo_close\n");

	if (c->keyfile != default_keyfile)
		free(c->keyfile);
	if (c->macfile != NULL)
		free(c->macfile);
	return (0);
}

